Haraka is throwing errors when the client sends QUIT too early

### Haraka version

2.7.2
### Observed behaviour

I occasionally see clients send 'QUIT' commands  before we send '250 Message Queued' replies.
In those situations Haraka throws:

`2016-04-08T13:54:53.00749 [ERROR] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] cmd_quit failed: Error: We are already running hooks! Fatal error!
2016-04-08T13:54:53.00778 [ERROR] [-] [core] undefinedError: We are already running hooks! Fatal error!0[ 'Error: We are already running hooks! Fatal error!',
2016-04-08T13:54:53.00779   '    at Object.plugins.run_hooks (/usr/local/lib/node_modules/Haraka/plugins.js:281:15)',
2016-04-08T13:54:53.00780   '    at Connection.cmd_quit (/usr/local/lib/node_modules/Haraka/connection.js:1193:13)',
2016-04-08T13:54:53.00781   '    at Connection.process_line (/usr/local/lib/node_modules/Haraka/connection.js:280:29)',
2016-04-08T13:54:53.00782   '    at Connection._process_data (/usr/local/lib/node_modules/Haraka/connection.js:423:18)',
2016-04-08T13:54:53.00783   '    at /usr/local/lib/node_modules/Haraka/connection.js:632:41',
2016-04-08T13:54:53.00783   '    at process._tickDomainCallback (node.js:486:13)' ]`
### Expected behaviour

1) Any ideas why the client is behaving this way? 
2) Any way to handle this more cleanly? In particular, I'd prefer to have the option to DENY these, because with 421 replies I keep getting this message sent (and the subsequent error) over and over.
### Steps to reproduce

Here is the last bit of data from the client before the error occurs:

```
...
2016-04-08T13:54:53.00168 [DATA] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] C: 0px; font-size: 0px; overflow: hidden; mso-hide: all"/></body>\r
2016-04-08T13:54:53.00181 [DATA] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] C: </html>\r
2016-04-08T13:54:53.00193 [DATA] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] C: <!-- /Wrapper -->\r
2016-04-08T13:54:53.00206 [DATA] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] C: ------=_NextPart_000_0186_01D1911A.E9F01E60--\r
2016-04-08T13:54:53.00220 [DATA] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] C: \r
2016-04-08T13:54:53.00233 [DATA] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] C: .\r
2016-04-08T13:54:53.00307 [DEBUG] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] running data_post hooks
2016-04-08T13:54:53.00319 [DEBUG] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] running data_post hook in hdmailer plugin
2016-04-08T13:54:53.00646 [PROTOCOL] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] C: QUIT state=1
2016-04-08T13:54:53.00656 [DEBUG] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] running quit hooks
2016-04-08T13:54:53.00749 [ERROR] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] cmd_quit failed: Error: We are already running hooks! Fatal error!
```

Thanks, let me know if there's any more info I can provide.

Dave

Hi Dave, nice to hear from you!

(for other Haraka devs reading this - I used to work at the company using
this, and wrote the original plugin here)

Knowing a bit about what goes on in hdmailer, it's related to handling
attachments and pausing/resuming is my guess.

We wouldn't process the QUIT from the client if the connection was paused.
At least I hope not - that code hasn't broken for my uses of it.

I thought I had put enough code into hdmailer to support this - I guess not.

You can see what is required by looking into the attachment plugin:
https://github.com/haraka/Haraka/blob/master/plugins/attachment.js - look
for where we call pause and resume. Check against those places in hdmailer
and see if you're missing something. The basic principal is that you're
calling resume() too soon. Add logging to the plugin to see if you're
calling resume before the processing of attachments ends.

Alternatively just contact me personally (you have my email) and I'll take
a poke at the code - it might be a straightforward fix.

As a side note, I'd love to improve all of this hackery. But it's a big
back-burner ticket in Haraka. Unfortunately due to attachments using
streams, we have yet to figure out a way to make this transparent to
plugins. Maybe for 3.0.

On Fri, Apr 8, 2016 at 6:21 PM, Dave Stibrany <notifications@github.com>
wrote:

> Haraka version
>
> 2.7.2
> Observed behaviour
>
> I occasionally see clients send 'QUIT' commands before we send '250
> Message Queued' replies.
> In those situations Haraka throws:
>
> 2016-04-08T13:54:53.00749 [ERROR] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1]
> [core] cmd_quit failed: Error: We are already running hooks! Fatal error!
> 2016-04-08T13:54:53.00778 [ERROR] [-] [core] undefinedError: We are
> already running hooks! Fatal error!0[ 'Error: We are already running hooks!
> Fatal error!',
> 2016-04-08T13:54:53.00779 ' at Object.plugins.run_hooks
> (/usr/local/lib/node_modules/Haraka/plugins.js:281:15)',
> 2016-04-08T13:54:53.00780 ' at Connection.cmd_quit
> (/usr/local/lib/node_modules/Haraka/connection.js:1193:13)',
> 2016-04-08T13:54:53.00781 ' at Connection.process_line
> (/usr/local/lib/node_modules/Haraka/connection.js:280:29)',
> 2016-04-08T13:54:53.00782 ' at Connection._process_data
> (/usr/local/lib/node_modules/Haraka/connection.js:423:18)',
> 2016-04-08T13:54:53.00783 ' at
> /usr/local/lib/node_modules/Haraka/connection.js:632:41',
> 2016-04-08T13:54:53.00783 ' at process._tickDomainCallback
> (node.js:486:13)' ]
> Expected behaviour
>
> 1) Any ideas why the client is behaving this way?
> 2) Any way to handle this more cleanly? In particular, I'd prefer to have
> the option to DENY these, because with 421 replies I keep getting this
> message sent (and the subsequent error) over and over.
> Steps to reproduce
>
> Here is the last bit of data from the client before the error occurs:
>
> ...
> 2016-04-08T13:54:53.00168 [DATA] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] C: 0px; font-size: 0px; overflow: hidden; mso-hide: all"/></body>\r
> 2016-04-08T13:54:53.00181 [DATA] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] C: </html>\r
> 2016-04-08T13:54:53.00193 [DATA] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] C: <!-- /Wrapper -->\r
> 2016-04-08T13:54:53.00206 [DATA] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] C: ------=_NextPart_000_0186_01D1911A.E9F01E60--\r
> 2016-04-08T13:54:53.00220 [DATA] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] C: \r
> 2016-04-08T13:54:53.00233 [DATA] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] C: .\r
> 2016-04-08T13:54:53.00307 [DEBUG] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] running data_post hooks
> 2016-04-08T13:54:53.00319 [DEBUG] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] running data_post hook in hdmailer plugin
> 2016-04-08T13:54:53.00646 [PROTOCOL] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] C: QUIT state=1
> 2016-04-08T13:54:53.00656 [DEBUG] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] running quit hooks
> 2016-04-08T13:54:53.00749 [ERROR] [EF24362D-72DC-4DBB-8B60-00DAD4A12BEF.1] [core] cmd_quit failed: Error: We are already running hooks! Fatal error!
>
> Thanks, let me know if there's any more info I can provide.
>
> Dave
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/haraka/Haraka/issues/1410>
>

Not a hard crash, just lots of errors:
```
[ERROR] [BA8E40D0-3D19-4233-8C54-386388B00007.1] [core] cmd_quit failed: Error: We are already running hooks! Fatal error!
[ERROR] [-] [core] undefinedError: We are already running hooks! Fatal error!0[ 'Error: We are already running hooks! Fatal error!',
  '    at Object.plugins.run_hooks (/home/vagrant/.nvm/versions/node/v4.6.0/lib/node_modules/Haraka/plugins.js:396:15)',
  '    at Connection.cmd_quit (/home/vagrant/.nvm/versions/node/v4.6.0/lib/node_modules/Haraka/connection.js:1279:13)',
  '    at Connection.process_line (/home/vagrant/.nvm/versions/node/v4.6.0/lib/node_modules/Haraka/connection.js:319:29)',
  '    at Connection._process_data (/home/vagrant/.nvm/versions/node/v4.6.0/lib/node_modules/Haraka/connection.js:464:18)',
  '    at Immediate._onImmediate (/home/vagrant/.nvm/versions/node/v4.6.0/lib/node_modules/Haraka/connection.js:680:37)',
  '    at processImmediate [as _immediateCallback] (timers.js:383:17)' ]
```

and then we send a 421 Internal Server Error

OK, but that error also shouldn't occur. I'll have a look anyway.

On Tue, Mar 21, 2017 at 1:28 PM, Dave Stibrany <notifications@github.com>
wrote:

> Not a hard crash, just lots of errors:
>
> [ERROR] [BA8E40D0-3D19-4233-8C54-386388B00007.1] [core] cmd_quit failed: Error: We are already running hooks! Fatal error!
> [ERROR] [-] [core] undefinedError: We are already running hooks! Fatal error!0[ 'Error: We are already running hooks! Fatal error!',
>   '    at Object.plugins.run_hooks (/home/vagrant/.nvm/versions/node/v4.6.0/lib/node_modules/Haraka/plugins.js:396:15)',
>   '    at Connection.cmd_quit (/home/vagrant/.nvm/versions/node/v4.6.0/lib/node_modules/Haraka/connection.js:1279:13)',
>   '    at Connection.process_line (/home/vagrant/.nvm/versions/node/v4.6.0/lib/node_modules/Haraka/connection.js:319:29)',
>   '    at Connection._process_data (/home/vagrant/.nvm/versions/node/v4.6.0/lib/node_modules/Haraka/connection.js:464:18)',
>   '    at Immediate._onImmediate (/home/vagrant/.nvm/versions/node/v4.6.0/lib/node_modules/Haraka/connection.js:680:37)',
>   '    at processImmediate [as _immediateCallback] (timers.js:383:17)' ]
>
> and then we send a 421 Internal Server Error
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/haraka/Haraka/issues/1410#issuecomment-288155666>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAobY9lp0wFNP8qa7ijJPLKzIPIlziPlks5roAjFgaJpZM4IDbym>
> .
>

Is it possible that the client is doing something bizarre? This issue does not happen with > 99% of our users, but happens consistently for those ones.  Here's an example, the attachment processing took < 1s but the client still appears to be sending the QUIT early, or has already sent it.

```
2017-04-10T17:41:58.22501 [PROTOCOL] [7AFDF64E-85E2-44B7-9819-F2A7EC6F5ADC] [core] C: EHLO server.example.com state=1
... rest of the message
2017-04-10T17:41:58.70248 [DATA] [7AFDF64E-85E2-44B7-9819-F2A7EC6F5ADC.1] [core] C: .\r
2017-04-10T17:41:58.70288 [DEBUG] [7AFDF64E-85E2-44B7-9819-F2A7EC6F5ADC.1] [core] running data_post hooks
2017-04-10T17:41:58.70306 [DEBUG] [7AFDF64E-85E2-44B7-9819-F2A7EC6F5ADC.1] [core] running data_post hook in hdmailer plugin
2017-04-10T17:41:58.70384 [PROTOCOL] [7AFDF64E-85E2-44B7-9819-F2A7EC6F5ADC.1] [core] C: QUIT state=1
2017-04-10T17:41:58.70396 [DEBUG] [7AFDF64E-85E2-44B7-9819-F2A7EC6F5ADC.1] [core] running quit hooks
2017-04-10T17:41:58.70464 [ERROR] [7AFDF64E-85E2-44B7-9819-F2A7EC6F5ADC.1] [core] cmd_quit failed: Error: We are already running hooks! Fatal error!
2017-04-10T17:41:58.70703 [PROTOCOL] [7AFDF64E-85E2-44B7-9819-F2A7EC6F5ADC.1] [core] S: 421 Internal Server Error
2017-04-10T17:41:58.70734 [DEBUG] [7AFDF64E-85E2-44B7-9819-F2A7EC6F5ADC.1] [core] running reset_transaction hooks
2017-04-10T17:41:58.70747 [DEBUG] [7AFDF64E-85E2-44B7-9819-F2A7EC6F5ADC.1] [core] running disconnect hooks
2017-04-10T17:41:58.70761 [NOTICE] [7AFDF64E-85E2-44B7-9819-F2A7EC6F5ADC.1] [core] disconnect ip=x.x.x.222 rdns="unallocated-static.rogers.com" helo="server.example.com" relay=N early=N esmtp=Y tls=N pipe=Y errors=0 txns=1 rcpts=1/0/0 msgs=0/0/0 bytes=120808 lr="421 Internal Server Error" time=0.572
```

--- a/connection.js
+++ b/connection.js
@@ -1605,9 +1605,7 @@ Connection.prototype.accumulate_data = function (line) {
 
 Connection.prototype.data_done = function () {
     var self = this;
-    this.state = states.STATE_CMD;
     this.pause();
-    // this.state = states.STATE_PAUSE;
     this.totalbytes += this.transaction.data_bytes;
 
     // Check message size limit

--- a/plugins/attachment.js
+++ b/plugins/attachment.js
@@ -496,10 +496,9 @@ exports.check_items_against_regexps = function (items, regexps) {
 };
 
 exports.wait_for_attachment_hooks = function (next, connection) {
-    this.loginfo("Checking if attachment hooks are running");
     var txn = connection.transaction;
     if (txn.notes.attachment_count > 0) {
-        this.loginfo("We still have attachment hooks running");
+        // this.loginfo("We still have attachment hooks running");
         txn.notes.attachment_next = next;
     }
     else {

--- a/plugins/attachment.js
+++ b/plugins/attachment.js
@@ -293,6 +293,8 @@ exports.start_attachment = function (connection, ctype, filename, body, stream) 
         bytes += data.length;
     });
     stream.once('end', function () {
+        stream.pause();
+
         digest = md5.digest('hex');
         connection.loginfo(plugin, 'file="' + filename + '" ctype="' + ctype +
                                    '" md5=' + digest + ' bytes=' + bytes);
@@ -325,6 +327,7 @@ exports.start_attachment = function (connection, ctype, filename, body, stream) 
                     connection.logdebug(plugin, 'unlinked: ' + fn);
                 });
             });
+            stream.resume();
         }
         if (err) {
             txn.notes.attachment_result = [ DENYSOFT, err.message ];
@@ -350,6 +353,7 @@ exports.start_attachment = function (connection, ctype, filename, body, stream) 
         });
         ws.on('close', function () {
             connection.logdebug(plugin, 'end of stream reached');
+            connection.pause();
             plugin.unarchive_recursive(connection, fn, filename, function (error, files) {
                 txn.notes.attachment_count--;
                 cleanup();
@@ -375,6 +379,7 @@ exports.start_attachment = function (connection, ctype, filename, body, stream) 
                 else {
                     txn.notes.attachment_archive_files = txn.notes.attachment_archive_files.concat(files);
                 }
+                connection.resume();
                 return next();
             });
         });
@@ -491,9 +496,10 @@ exports.check_items_against_regexps = function (items, regexps) {
 };
 
 exports.wait_for_attachment_hooks = function (next, connection) {
+    this.loginfo("Checking if attachment hooks are running");
     var txn = connection.transaction;
     if (txn.notes.attachment_count > 0) {
-        // We still have attachment hooks running
+        this.loginfo("We still have attachment hooks running");
         txn.notes.attachment_next = next;
     }
     else {

